Bugfix: replace addClass with getClass in UCC listener

CoralMalachi opened the pull request:
Currently, when trying to send a 1905 message of type
`CHANNEL_SELECTION_REQUEST_MESSAGE` or `COMBINED_INFRASTRUCTURE_METRICS_MESSAGE`,
the UCC listener replay status running and stack there.

The reason for this is the using of addClass in `beerocks_ucc_listener::create_cmdu`
was not replaced with the new `getClass` method as it should.

As a result, the UCC listener got stuck in an infinite loop, since the
parse of the TLVs already happened, which caused a timeout error.

In order to solve it, both while loops are removed and replaced with two
simple calls to `getClass` method.
Also, `tlv_type` is removed since it is not needed anymore.

Fixes #622 (checked on testbed)

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
common: ucc listener: replace addClass with getClass

Currently, when trying to send a 1905 message of type
CHANNEL_SELECTION_REQUEST_MESSAGE or COMBINED_INFRASTRUCTURE_METRICS_MESSAGE,
the UCC listener replay status running and stack there.

The reason for this is the using of addClass in `beerocks_ucc_listener::create_cmdu`
was not replaced with the new getClass method as it should.

As a result, the UCC listener got stuck in an infinite loop, since the
parse of the TLVs already happened, which caused to a timeout error.

In order to solve it, both while loops are removed and replaced with two
simple calls to getClass method.
Also, `tlv_type` is removed since it is not needed anymore.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

tomereli reviewed (approved):
Let's also make addClass protected and add a wrapper for it only in CmduMessageTx. On RX it should never be used

arnout commented on common/beerocks/bcl/source/beerocks_ucc_listener.cpp:
```diff
@@ -277,23 +275,18 @@ bool beerocks_ucc_listener::create_cmdu(ieee1905_1::CmduMessageTx &cmdu_tx,
-                }
-            }
-            tlv_type = cmdu_tx.getNextTlvType();
+        auto tlvChannelPreference = cmdu_tx.getClass<wfa_map::tlvChannelPreference>();
+        if (!tlvChannelPreference) {
+            LOG(ERROR) << "getClass<tlvChannelPreference> failed";
+            return false;
         }
+
+        auto tlvTransmitPowerLimit = cmdu_tx.getClass<wfa_map::tlvTransmitPowerLimit>();
+        if (!tlvTransmitPowerLimit) {
+            LOG(ERROR) << "getClass<tlvTransmitPowerLimit> failed";
+            return false;
```
I don't think these checks are needed anymore. We only did it before to make sure the structures were parsed properly. Now the complete parsing is done as part of load(), so the addClass calls simply aren't needed. It's up to the functions that prepare the certification buffer to make sure that it contains the appropriate CMDUs.

So basically, I think that this entire function can be simplified to:

```
bool beerocks_ucc_listener::create_cmdu(ieee1905_1::CmduMessageTx &cmdu_tx,
                                        ieee1905_1::eMessageType message_type,
                                        std::string &err_string)
{
    switch (message_type) {
    case ieee1905_1::eMessageType::CHANNEL_SELECTION_REQUEST_MESSAGE:
    case ieee1905_1::eMessageType::COMBINED_INFRASTRUCTURE_METRICS_MESSAGE: {

        auto cmdu_header = cmdu_tx.load();
        if (!cmdu_header) {
            LOG(ERROR) << "load cmdu has failed";
            err_string = err_internal;
            return false;
        }
        // Force mid
        cmdu_header->message_id() = g_mid;
        
        return true;
    }

    default:
        // If message type handler is not implemented, then return false with empty error, so the
        // caller would know that it was not handled.
        err_string = std::string();
        return false;
    }
}
```

CoralMalachi replied:
I think we should use `getClass` here.
The reason for this is that we call `create_cmdu` method at line 790 in `beerocks_ucc_listener.cpp` to check if the current  CMDU message has a preprepared buffer which should be loaded, otherwise, we need to create it manually and use TLV data from the command if exists.
if we remove those checks we'll never know in which situation are we.

arnout replied:
I disagree. But since this whole certification buffer thing is a horrible hack anyway, let's not discuss it further.

morantr replied:
@CoralMalachi, Arnout has suggested a correct implementation IMO. We used `AddClass` before, to add the classes into the CMDU vector so it will be swapped on `finalize()`. Now that the data on the buffer is always swapped there is no need to do it anymore, so getting the classes is useless now. The only thing that needs to be done is to modify the `MID` on the CMDU header.

If I'm not mistaken, the data on the CMDU header we get is already swapped, so I think that we need to override its value with swapped value:
 `cmdu_header->message_id() = swap_16(g_mid);`

tomereli replied:
One important thing to note is that any `getClass` (except for cmdu header) will return `nullptr` after `cmdu_tx.load()` - so these checks are definitely not needed since they will always result with the same outcome.

CoralMalachi replied:
As @tomereli said since the `getClass` always returns `nullptr` we pass the test by accident. 
The `CHANNEL_SELECTION_REQUEST_MESSAGE` CMDU should be loaded from the certification buffer, but instead, it is created manually and use TLVs from the command (which do not exist).

To reveal this kind of bug, when a CMDU is not created from the certification buffer when it should be, update the test_flows to validate it in the last commit.

Commit:
common: ucc listener: add logs

Preparative commit.

When we are in certficiation mode, it test flow 5.5.1 (Channel
Selection), we should send the Channel Selection CMDU only when we the
UCC asked us to.

Since the test_flows script illustrates the case we are in certification
mode, in the channel selection test we should validate that the CMDU is
loaded from the buffer and that we did not try to create it manually in
the UCC listener.

For that reason, after the CMDU in the certification buffer is loaded
successfully we add a print to notice it. In the next commit, test
channel_selection will search for this log message and fail the test
if did not find it.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tools: docker: update channel selection test (fails the test)

Currently, after the change of the first commit, we pass all the tests
even though test channel selection, for example, should fail.

The reason for this is that when we use `cmdu_tx.load()` in
`beerocks_ucc_listener::create_cmdu` method, does not parse the entire
buffer, but add the `cCmduHeader` only.

As a result, each call to `getClass` will return `nullptr`, and the
function returns false.

That way, we accidentally pass the test, since after
`beerocks_ucc_listener::create_cmdu` returned false, we assume that the
CMDU needs to be created manually and not to be loaded from the preprepared
buffer as it should be.

To reveal such a bug, update the `test_channel_selection` to search
for the log message which indicates that the CMDU was loaded successfully
from the preprepared buffer.

The same update will be added for the `test_combined_infra_metrics` in a
separate PR.

Please note that this update fails the channel selection test as it
should.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>
